### PR TITLE
Related stories

### PR DESCRIPTION
**Pull request by brunohkbx**
Display Tag below the project name in settings page

tag

Show related stories
This commit add the functionality to all projects belonging to the same tag group share to each other its release stories.
These stories of type release will be shown in the column done, in progress and backlog.

related stories

TO FIX:
All stories are positioned according to it value in the position field.
This causes a misbehavior that makes the stories of a given project display into a different iteration in another project of the same group tag.

PS : 
This PR is dependent from this other https://github.com/Codeminer42/cm42-central-support/pull/4

### CODE REVIEW

File: app/models/project.rb [hunk 1]
```diff
@@ -27,6 +27,10 @@ class Project < ActiveRecord::Base
 
   scope :joinable_except, -> (project_ids) { joinable.where.not(id: project_ids) }
 
+  def tag_group_attributes
+    tag_group&.attributes
```
Comment (adbatista) [inline app/models/project.rb]: you can use delegate here

File: app/assets/javascripts/models/story.js [hunk 2]
```diff
@@ -153,19 +152,28 @@ var Story = module.exports = Backbone.Model.extend({
...
-        return ["accept", "reject"];
-      case 'rejected':
-        return ["restart"];
-      case 'accepted':
-        return [];
-      default:
-        return ["start"];
+    var storyType = this.get('story_type');
+
+    if (storyType != 'related')
```
Comment (eddiecode) [inline app/assets/javascripts/models/story.js]: You don't need the else here, you can rewrite this logic to something like this:

```
if (storyType != 'related') {
  return [];
}

switch...
```

File: app/assets/javascripts/views/story_view.js [hunk 3]
```diff
@@ -591,16 +595,26 @@ module.exports = FormView.extend({
   },
 
   setClassName: function() {
+
```
Comment (marclerodrigues) [inline app/assets/javascripts/views/story_view.js]: remove this line. :nail_care:

File: app/assets/javascripts/models/story.js [hunk 4]
```diff
@@ -34,7 +34,7 @@ var Story = module.exports = Backbone.Model.extend({
   },
 
   setReadonly: function() {
-    if(this.get('state') === 'accepted' && this.get('accepted_at') !== undefined)
+    if(this.get('state') === 'accepted' && this.get('accepted_at') !== undefined || this.get('story_type') === 'related')
```
Comment (eddiecode) [inline app/assets/javascripts/models/story.js]: This line is too long, maybe you could break it into multiples lines

File: app/assets/javascripts/views/story_view.js (hunk 3 above)
Comment (eddiecode) [inline app/assets/javascripts/views/story_view.js]: Remove this line break.

File: app/queries/project_with_related_stories_query.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    related_projects = Project.where(tag_group: project.tag_group).where.not(id: project.id)
+
+    Story
+      .select(columns_for_related)
+      .where(project: related_projects, story_type: 'release')
+      .where.not(state: 'unscheduled')
+  end
+
+  def columns_for_related
+   columns << "'related' AS story_type"
```
Comment (marclerodrigues) [inline app/queries/project_with_related_stories_query.rb]: What do you think about moving these values for columns to constants?

File: app/queries/project_with_related_stories_query.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
+class ProjectWithRelatedStoriesQuery
```
Comment (eddiecode) [inline app/queries/project_with_related_stories_query.rb]: You need to write tests for this query object

File: app/queries/project_with_related_stories_query.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    relation.with_dependencies.order('updated_at DESC')
+  end
+
+  private
+
+  attr_reader :project, :relation
+
+  def related_stories
+    related_projects = Project.where(tag_group: project.tag_group).where.not(id: project.id)
```
Comment (marclerodrigues) [inline app/queries/project_with_related_stories_query.rb]: What about creating a new scope to deal with this?
Maybe something like:
```ruby
scope :related_projects, -> (id, tag_group) { where(tag_group: tag_group).where.not(id: id) }
```

File: app/views/projects/edit.html.erb [hunk 8]
```diff
@@ -1,7 +1,13 @@
 <% content_for :title_bar do %>
   <ul class="nav navbar-nav">
     <li>
-      <%= link_to @project.name, @project %>
+      <%= link_to @project.name, @project, { class: "project-tag" } %>
+      <% unless @project.tag_group.blank? %>
```
Comment (eddiecode) [inline app/views/projects/edit.html.erb]: I think you could use `if` instead of `unless`.

```
<% if @project.tag_group.present? %>
```

Comment (eddiecode) [inline app/views/projects/edit.html.erb]: You're using this code into other files, It's a better idea to create a partial for it.

File: app/models/story.rb [hunk 9]
```diff
@@ -62,6 +62,8 @@ def readonly?
     !accepted_at_changed? && accepted_at.present?
   end
 
+  scope :with_dependencies, -> { includes(:notes, :tasks, :document_files) }
```
Comment (marclerodrigues) [inline app/models/story.rb]: Add specs to this scope.

File: app/models/tag_group.rb [hunk 10]
```diff
@@ -2,6 +2,8 @@ class TagGroup < ActiveRecord::Base
   has_many :projects
   belongs_to :team
 
+  before_save :generate_foreground_color
```
Comment (marclerodrigues) [inline app/models/tag_group.rb]: Missing specs.

File: app/queries/project_with_related_stories_query.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    Story
```
Comment (marclerodrigues) [inline app/queries/project_with_related_stories_query.rb]: What do you think about moving this to a PORO or new scope?

File: app/assets/javascripts/models/story.js [hunk 12]
```diff
@@ -153,20 +155,24 @@ var Story = module.exports = Backbone.Model.extend({
...
-      case 'delivered':
-        return ["accept", "reject"];
-      case 'rejected':
-        return ["restart"];
-      case 'accepted':
-        return [];
-      default:
-        return ["start"];
-    }
+    var storyType = this.get('story_type');
```
Comment (talyssonoc) [inline app/assets/javascripts/models/story.js]: I think you're not using this variable.

**Comment by andrerpbts** [comment]
Closing this for now. This changes is very outdated, and the solution seems to be different right now.

**Pull request closed without merging**